### PR TITLE
Serialize whether VarDecls are top-level

### PR DESCRIPTION
**Pull request by adrian-prantl**
The current way that VarDecl::isLazilyInitializedGlobal() is implemented does
not work in the debugger, since the DeclContext of all VarDecls are deserialized
Swift modules. By adding a bit to the VarDecl we can recover the fact that a
VarDecl was in fact a global even in the debugger.

rdar://problem/58939370

### CODE REVIEW

**Comment by adrian-prantl** [comment]
Any pointers for how to test this would be appreciated.

**Review by xedin — APPROVED** [review]

**Comment by xedin** [comment]
I can pull this change into my branch and run full suite test to see whether it fixes the problem.

**Comment by xedin** [comment]
@adrian-prantl I have `am`ed your commit to https://github.com/apple/swift/pull/29024 and started full test.

File: lib/SILGen/SILGenGlobalVariable.cpp [hunk 1]
```diff
@@ -69,7 +69,8 @@ ManagedValue
 SILGenFunction::emitGlobalVariableRef(SILLocation loc, VarDecl *var) {
   assert(!VarLocs.count(var));
 
-  if (var->isLazilyInitializedGlobal()) {
+  // In the debugger, no variables are lazy initialized.
+  if (!SGM.SwiftModule->getDebugClient() && var->isLazilyInitializedGlobal()) {
```
Comment (adrian-prantl) [inline lib/SILGen/SILGenGlobalVariable.cpp]: Sorry, I accidentally left this is.

Comment (xedin) [reply lib/SILGen/SILGenGlobalVariable.cpp]: No worries, I can re-apply and test :)

**Comment by xedin** [comment]
I have re-applied the changes and restarted testing.

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/llvm-project/pull/674
@swift-ci test

**Comment by adrian-prantl** [comment]
@xedin You'll also need https://github.com/apple/llvm-project/pull/674 for the LLDB tests to succeed. This PR tests whether it also works *without* your PR. If yes I will land this, cherry-pick both to master-rebranch and master-next and then you should be able to merge your PR.

**Comment by xedin** [comment]
Sounds good! Going to keep an eye on this PR.

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/llvm-project/pull/674
@swift-ci test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/llvm-project/pull/674
@swift-ci smoke test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/llvm-project/pull/674
@swift-ci smoke test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/llvm-project/pull/674
@swift-ci test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/llvm-project/pull/674
@swift-ci test

File: lib/AST/Decl.cpp [hunk 2]
```diff
@@ -5410,11 +5411,7 @@ bool VarDecl::isLazilyInitializedGlobal() const {
 
   // Top-level global variables in the main source file and in the REPL are not
```
Comment (slavapestov) [inline lib/AST/Decl.cpp]: Are the above checks (hasClangNode() and isDebuggerVar()) needed at all then? The bit should not be set in those cases either.

**Review by slavapestov — COMMENTED** [review]
This looks good, but it appears you've cherry-picked @xedin's changes into this PR too. Can we land the two sets of changes separately?

**Comment by xedin** [comment]
@slavapestov We actually going to do it vice versa - I have cherry-picked Adrian's changes to my PR to make sure everything works correctly, currently waiting on full performance suite to complete and we are going to land my PR together with llvm side.

**Comment by slavapestov** [comment]
@xedin Ok, sure. In that case, can you simplify VarDecl::isLazilyInitializedGlobal() in your PR then?

**Comment by xedin** [comment]
@slavapestov Sure! I'm going to let full test finish and commit the simplification to my PR.

**Comment by adrian-prantl** [comment]
Abandoning this in favor of @xedin's all-in-one PR.

**Pull request closed without merging**